src/beatree/leaf/free_list.rs

```diff
@@ -0,0 +1,293 @@
+    ///
+    /// When a new portion is required, the page number is taken from the previous head of the list.
+    ///
+    /// There will be some fragmentation in the old head of the free list,
+    /// for example:
+    ///   * Free List: 1, 7, 9, 2, X (max 5 PageNumbers per portion)
+    ///   * Appending: 10, 18
+    ///   * Result: 1, 7, 9, 10, X | 18
+    ///
+    /// 18 is allocated into a new portion with page number 2,
+    /// and the previous head will be rewritten with a little fragmentation inside (4 bytes).
+    ///
+    /// If the list is empty, page numbers are taken from the nonce.
+    ///
+    /// Returns two vectors containing the pages to be written into the store
+    pub fn encode(
+        &mut self,
+        mut to_append: Vec<PageNumber>,
+        nonce: &mut PageNumber,
+        max_nonce: PageNumber,
+    ) -> FreeListAppendOutput {
+        // append the released free list pages
+        to_append.extend(std::mem::take(&mut self.released_portions));
+
+        let new_pns_len = to_append.len();
+
+        // max number of pages required to be allocated
+        let pages_to_allcoate = (new_pns_len as f64 / MAX_PNS_PER_FREE_PAGE as f64).ceil() as usize;
+
+        let additional = if self.head.is_none() { 0 } else { 1 };
+        let mut free_list_pages_pns = (0..pages_to_allcoate + additional)
+            .map(|_| match self.pop() {
+                Some(pn) => pn,
+                None => {
+                    let pn = *nonce;
+                    nonce.0 += 1;
+                    pn
+                }
+            })
+            .collect::<Vec<_>>()
+            .into_iter();
+
+        let mut pns_iter = to_append.into_iter();
+        let mut inner_frag = false;
+
+        // if the head is not empty, then the previous pop
+        // did not empty the free list, thus there will be some free space
+        // left in the head.
+        if let Some((head_pn, head)) = &mut self.head {
+            // let's change the page_number of the current head to avoid overwriting it
+            *head_pn = free_list_pages_pns.next().unwrap();
+
+            let free_space = MAX_PNS_PER_FREE_PAGE - head.len();
+
+            // if this holds then we will have to deal with a PageNumber fragmentation in the new
+            // head
+            if free_space + (pages_to_allcoate - 1) * MAX_PNS_PER_FREE_PAGE == new_pns_len {
+                head.extend(pns_iter.by_ref().take(free_space - 1));
+                inner_frag = true;
+            }
+        }
+
+        let mut create_pages = vec![];
+
+        for pn in pns_iter {
+            let maybe_encoded_head = self.push(pn, &mut free_list_pages_pns, inner_frag);
+            inner_frag = false;
+
+            if let Some(encoded) = maybe_encoded_head {
+                create_pages.push(encoded);
+            }
+        }
+
+        FreeListAppendOutput::create_output(create_pages, max_nonce)
```

Comment:
doc update with:

```rust
/// Pages numbers are updated during batches of commits by taking free pages from the free list
/// or adding new ones to the list.
///
/// FreeList Pages are stored in the LeafStore using a Copy-On-Write (COW) approach.
///
/// This function considers every step taken since the last call to this function,
/// creating an encoded version of the updated free list state. (ready to be saved into the LeafStore)
///
/// Possible operations on a FreeList include removing a free page or adding a new one.
/// These operations, while creating the encoded format, lead to multiple intermediate steps:
///    1. Removing pages consumes the list.
///    2. Adding pages appends them to the list's head.
///    3. Adding new pages require allocating a new page for the list's head.
///       This page should not overwrite pages in the leaf store and
///       should be taken from the head at the end of step one.
///
/// Shifting page numbers to the old head involves moving all page numbers added at step two.
/// Generally, this is not an issue, for example using FreeList pages that can store up to five numbers:
///
/// The current head is composed of: [1, 7, 9, 2, 5].
///
/// The performed operations include one pop and pushing five elements ([6, 10, 12, 15, 20]).
///
/// As a result, 5 is popped out, leaving space for only 6 in the head.
/// Consequently, a new head is created to store [10, 12, 15, 20].
///
/// To accommodate the new head and overwrite the old one, two new page numbers are required.
/// These are taken from the head at the end of stage 1, after removing all values.
///
/// In this case, 2 and 9 will be used. With two additional free spaces in the previous head,
/// values from the new head will shift to occupy all spots in the old head. The final result is:
///
/// [(page_number: 2, free_list:[1, 7, 6, 10, 12]), (page_number: 5, free_list:[15, 20])]
///
/// There is a special case where shifting values to the left can result in the new head becoming empty.
/// For example, starting from the previous state, if we pop 5 elements and push only [10, 12, 15]:
///
/// - Head after step 1: [1, 7, 9, 2]
/// - We need to push 3 new elements, so 10 will be the last element in the head.
/// - Elements 2 and 9 are removed from the old head and used as new page numbers for both the old and new heads.
/// - The new head should store 12 and 15.
/// - The removal of 2 and 9 causes a shift, resulting in all elements fitting into the old head: [1, 7, 10, 12, 15]
///   with a page number of 2 but unused page number 9.
/// - To maintain the LIFO property and avoid wasting page number 9, a small fragmentation is created.
///   The old head will then contain only 4 elements, and the new head will contain one element, resulting in:
///  
/// [(Page number: 2, Free list: [1, 7, 10, 12]), (Page number: 9, Free list: [15])]
```